test/replication-luatest/election_pre_vote_test.lua

```diff
@@ -0,0 +1,114 @@
+local t = require('luatest')
+local cluster = require('test.luatest_helpers.cluster')
+local server = require('test.luatest_helpers.server')
+
+local g = t.group('pre-vote')
+
+local REPLICATION_TIMEOUT = 0.1
+local DEATH_TIMEOUT = 5 * REPLICATION_TIMEOUT
+
+--
+-- gh-6654: pre-vote.
+-- The nodes shouldn't start elections when the elections would be either
+-- disruptive (there's an active leader somewhere) or pointless (the node
+-- doesn't have a quorum of peers).
+--
+g.before_all(function()
+    g.cluster = cluster:new({})
+    g.cfg = {
+        replication = {
+            server.build_instance_uri('node1'),
+            server.build_instance_uri('node2'),
+            server.build_instance_uri('node3'),
+        },
+        election_mode = 'candidate',
+        replication_timeout = REPLICATION_TIMEOUT,
+    }
+    g.node1 = g.cluster:build_and_add_server({alias = 'node1', box_cfg = g.cfg})
+    g.node2 = g.cluster:build_and_add_server({alias = 'node2', box_cfg = g.cfg})
+    g.node3 = g.cluster:build_and_add_server({alias = 'node3', box_cfg = g.cfg})
+    g.cluster:start()
+end)
+
+g.before_each(function()
+    g.leader = g.cluster:get_leader()
+    t.assert(g.leader ~= nil, 'Cluster elected a leader')
+    g.follower1 = g.node1 ~= g.leader and g.node1 or g.node2
+    g.follower2 = g.node3 ~= g.leader and g.node3 or g.node2
+end)
+
+g.after_each(function()
+    g.leader = nil
+    g.follower1 = nil
+    g.follower2 = nil
+end)
+
+--
+-- Test that the node doesn't start elections in the following conditions:
+--  - the cluster has a leader
+--  - the node doesn't see the leader directly
+--  - the node has a quorum of connections to remote peers (excluding leader)
+--
+g.test_no_direct_connection = function(g)
+    local term = g.follower1:exec(function() return box.info.election.term end)
+    g.follower1:exec(function(cfg) box.cfg(cfg) end,
+                     {{
+                         replication = {
+                            server.build_instance_uri(g.follower1.alias),
+                            server.build_instance_uri(g.follower2.alias),
+                         }
+                     }})
+    t.helpers.retrying({}, function()
+        g.follower1:exec(function(t)
+            require('luatest').assert(box.info.election.leader_idle >= t)
+        end, {DEATH_TIMEOUT})
+    end)
+    t.assert_equals(
+        g.follower1:exec(function() return box.info.election.term end), term,
```

Comment:
Something like that you might like more, up to you:
```Diff
@@ -43,6 +43,10 @@ g.after_each(function()
     g.follower2 = nil
 end)
 
+local function get_election_term()
+    return box.info.election.term
+end
+
 --
 -- Test that the node doesn't start elections in the following conditions:
 --  - the cluster has a leader
@@ -50,7 +54,7 @@ end)
 --  - the node has a quorum of connections to remote peers (excluding leader)
 --
 g.test_no_direct_connection = function(g)
-    local term = g.follower1:exec(function() return box.info.election.term end)
+    local term = g.follower1:exec(get_election_term)
     g.follower1:exec(function(cfg) box.cfg(cfg) end,
                      {{
                          replication = {
@@ -64,12 +68,12 @@ g.test_no_direct_connection = function(g)
         end, {DEATH_TIMEOUT})
     end)
     t.assert_equals(
-        g.follower1:exec(function() return box.info.election.term end), term,
+        g.follower1:exec(get_election_term), term,
         'No elections when leader is seen through some proxy node')
     g.follower1:exec(function(cfg) box.cfg(cfg) end,
                      {{replication = g.cfg.replication}})
     t.assert_equals(
-        g.follower1:exec(function() return box.info.election.term end), term,
+        g.follower1:exec(get_election_term), term,
         'No elections after leader connection regain')
 
 end
@@ -82,14 +86,14 @@ end
 --
 g.test_no_quorum = function(g)
     g.follower1:exec(function() box.cfg{replication = ''} end)
-    local term = g.follower1:exec(function() return box.info.election.term end)
+    local term = g.follower1:exec(get_election_term)
     t.helpers.retrying({}, function()
         g.follower1:exec(function(t)
             require('luatest').assert(box.info.election.leader_idle >= t)
         end, {DEATH_TIMEOUT})
     end)
     t.assert_equals(
-        g.follower1:exec(function() return box.info.election.term end), term,
+        g.follower1:exec(get_election_term), term,
         'No elections when no leader and no quorum')
     g.follower1:exec(function(cfg) box.cfg(cfg) end,
                      {{
@@ -99,12 +103,12 @@ g.test_no_quorum = function(g)
                          }
                      }})
     t.assert_equals(
-        g.follower1:exec(function() return box.info.election.term end), term,
+        g.follower1:exec(get_election_term), term,
         'No elections after quorum regain')
     g.follower1:exec(function(cfg) box.cfg(cfg) end,
                      {replication = g.cfg.replication})
     t.assert_equals(
-        g.follower1:exec(function() return box.info.election.term end), term,
+        g.follower1:exec(get_election_term), term,
         'No elections after leader reconnect')
 end
```

Reply:
Applied, thanks!